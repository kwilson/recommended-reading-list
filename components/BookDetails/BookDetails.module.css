.container {
  display: grid;
  grid-template-columns: 1fr;
  grid-template-rows: auto;
  gap: 1rem;
  grid-template-areas:
    'title'
    'author'
    'isbn'
    'image'
    'otherBooks';
}

@media screen and (min-width: 760px) {
  .container {
    display: grid;
    grid-template-columns: 1fr 1fr;
    grid-template-rows: min-content min-content max-content min-content;
    gap: 1rem;
    grid-template-areas:
      'title title'
      'author author'
      'image isbn'
      'image otherBooks';
  }
}

.title {
  grid-area: title;
  margin-bottom: 0;
}
.image {
  grid-area: image;
}
.image img {
  border: solid 1px var(--border-color);
  height: auto;
  max-width: 100%;
}
.author {
  grid-area: author;
}
.isbn {
  grid-area: isbn;
  align-self: start;
}
.otherBooks {
  grid-area: otherBooks;
  align-self: start;
}
