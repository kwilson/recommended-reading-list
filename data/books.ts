import { parseIsbn, parseString, parseNumber } from '../utils/data-parser';

export interface IBookDataResponseItem {
  id: string;
  cover: string;
  isbn: string;
  title: string;
  subtitle: string;
  author: string;
  published: string;
  publisher?: string;
  pages?: number;
  description?: string;
  website?: string;
}

export interface IBookDataResponse {
  books: IBookDataResponseItem[];
}

export interface IBook {
  id: string;
  title: string;
  author: string;
  cover: string;
}

export interface IBookDetails extends IBook {
  isbn: string | null;
  subtitle: string;
  published: string;
  publisher: string | null;
  pages: number | null;
  description: string | null;
  website: string | null;
}

const endpoint = 'https://hokodo-frontend-interview.netlify.app/data.json';

async function getBooksData(): Promise<IBookDataResponse> {
  const data = await fetch(endpoint);
  const values: IBookDataResponse = await data.json();

  return values;
}

export const getBooks = async (): Promise<IBook[]> => {
  try {
    const data = await getBooksData();
    return data.books.map((input) => ({
      author: input.author,
      cover: input.cover,
      id: input.id,
      title: input.title,
    }));
  } catch (e) {
    return [];
  }
};

export const getBooksByAuthor = async (
  author: string,
  exludeIds: string[] = []
): Promise<IBook[]> => {
  try {
    const data = await getBooksData();
    const predicate = (book: IBook) =>
      book.author === author && !exludeIds.includes(book.id);

    return data.books.filter(predicate).map((input) => ({
      author: input.author,
      cover: input.cover,
      id: input.id,
      title: input.title,
    }));
  } catch (e) {
    return [];
  }
};

export const getBook = async (
  id: string | null
): Promise<IBookDetails | null> => {
  try {
    const { books } = await getBooksData();
    const match = books.find((book) => book.id === id);

    if (match) {
      return {
        author: match.author,
        cover: match.cover,
        description: parseString(match.description),
        id: match.id,
        isbn: parseIsbn(match.isbn),
        pages: parseNumber(match.pages),
        published: match.published,
        publisher: parseString(match.publisher),
        subtitle: match.subtitle,
        title: match.title,
        website: parseString(match.website),
      };
    }

    return null;
  } catch (e) {
    return null;
  }
};
