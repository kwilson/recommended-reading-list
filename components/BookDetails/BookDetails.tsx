import { FunctionComponent, useEffect, useState } from 'react';
import { IBook } from '../../data/books';
import { getBooksByAuthor } from '../../data/books';
import { LoaderContainer } from '../LoaderContainer';
import { BookList } from '../BookList';

import styles from './BookDetails.module.css';

interface IProps {
  id: string;
  title: string;
  author: string;
  isbn: string | null;
  cover: string;
}

export const BookDetails: FunctionComponent<IProps> = ({
  id,
  title,
  author,
  isbn,
  cover,
}) => {
  const [relatedBooks, setRelatedBooks] = useState<IBook[] | null>(null);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    setIsLoading(true);
    getBooksByAuthor(author, [id])
      .then(setRelatedBooks)
      .then(() => setIsLoading(false));
  }, [author, id]);

  return (
    <div className={styles.container} data-testid="book-details">
      <h1 className={styles.title} data-testid="title">
        {title}
      </h1>
      <div className={styles.author}>
        by <span data-testid="author">{author}</span>
      </div>
      <div className={styles.isbn}>
        ISBN: <span data-testid="isbn">{isbn ?? 'not available'}</span>
      </div>

      <div className={styles.image}>
        <img
          data-testid="image"
          className={styles.image}
          src={cover}
          alt={`${title} book cover`}
        />
      </div>

      <aside className={styles.otherBooks}>
        <h2>Other books by this author</h2>
        <LoaderContainer loading={isLoading}>
          {relatedBooks && <BookList books={relatedBooks} />}
        </LoaderContainer>
      </aside>
    </div>
  );
};
