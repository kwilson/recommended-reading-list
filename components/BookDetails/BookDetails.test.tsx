import { render, screen, waitFor } from '@testing-library/react';

import { getBooksByAuthor, IBook } from '../../data/books';
import { BookDetails } from './BookDetails';

jest.mock('../../data/books');
const mockGetBooksByAuthor = getBooksByAuthor as jest.Mock<
  ReturnType<typeof getBooksByAuthor>
>;

describe(`<${BookDetails.name} />`, () => {
  const props = {
    id: ' value',
    title: 'title value',
    author: 'author value',
    isbn: 'isbn value',
    cover: 'cover value',
  };

  it('renders content as expected', async () => {
    const promise = new Promise<IBook[]>((resolve) => {
      setTimeout(resolve, 1000);
    });
    mockGetBooksByAuthor.mockReturnValueOnce(promise);

    const { asFragment } = render(<BookDetails {...props} />);

    expect(screen.getByTestId('title').textContent).toEqual(props.title);
    expect(screen.getByTestId('author').textContent).toEqual(props.author);
    expect(screen.getByTestId('isbn').textContent).toEqual(props.isbn);
    expect(screen.getByTestId('image').getAttribute('src')).toEqual(
      props.cover
    );
    expect(screen.getByTestId('image').getAttribute('alt')).toEqual(
      expect.stringContaining(props.title)
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders content as expected when there is no ISBN', async () => {
    const promise = new Promise<IBook[]>((resolve) => {
      setTimeout(resolve, 1000);
    });
    mockGetBooksByAuthor.mockReturnValueOnce(promise);

    const { asFragment } = render(<BookDetails {...props} isbn={null} />);

    expect(screen.getByTestId('isbn').textContent).toEqual('not available');

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders as expected when loading', async () => {
    const promise = new Promise<IBook[]>((resolve) => {
      setTimeout(resolve, 1000);
    });
    mockGetBooksByAuthor.mockReturnValueOnce(promise);

    const { asFragment } = render(<BookDetails {...props} />);

    await waitFor(() => screen.getByTestId('loader'));
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders as expected when loaded with no related books', async () => {
    mockGetBooksByAuthor.mockResolvedValueOnce([]);

    const { asFragment } = render(<BookDetails {...props} />);

    await waitFor(() => screen.getByTestId('empty-book-list'));
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders as expected when loaded with related books', async () => {
    mockGetBooksByAuthor.mockResolvedValueOnce([
      {
        id: ' value 2',
        title: 'title value 2',
        author: 'author value',
        cover: 'cover value 2',
      },
    ]);

    const { asFragment } = render(<BookDetails {...props} />);

    await waitFor(() => screen.getByTestId('book-list-items'));
    expect(asFragment()).toMatchSnapshot();
  });
});
