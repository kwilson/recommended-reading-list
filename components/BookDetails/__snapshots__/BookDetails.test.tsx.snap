// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookDetails /> renders as expected when loaded with no related books 1`] = `
<DocumentFragment>
  <div
    class="container"
    data-testid="book-details"
  >
    <h1
      class="title"
      data-testid="title"
    >
      title value
    </h1>
    <div
      class="author"
    >
      by 
      <span
        data-testid="author"
      >
        author value
      </span>
    </div>
    <div
      class="isbn"
    >
      ISBN: 
      <span
        data-testid="isbn"
      >
        isbn value
      </span>
    </div>
    <div
      class="image"
    >
      <img
        alt="title value book cover"
        class="image"
        data-testid="image"
        src="cover value"
      />
    </div>
    <aside
      class="otherBooks"
    >
      <h2>
        Other books by this author
      </h2>
      <div
        class="container"
      >
        <div
          class="container"
        >
          <div
            data-testid="empty-book-list"
          >
            <p>
              Sorry, there are no books to display.
            </p>
          </div>
        </div>
      </div>
    </aside>
  </div>
</DocumentFragment>
`;

exports[`<BookDetails /> renders as expected when loaded with related books 1`] = `
<DocumentFragment>
  <div
    class="container"
    data-testid="book-details"
  >
    <h1
      class="title"
      data-testid="title"
    >
      title value
    </h1>
    <div
      class="author"
    >
      by 
      <span
        data-testid="author"
      >
        author value
      </span>
    </div>
    <div
      class="isbn"
    >
      ISBN: 
      <span
        data-testid="isbn"
      >
        isbn value
      </span>
    </div>
    <div
      class="image"
    >
      <img
        alt="title value book cover"
        class="image"
        data-testid="image"
        src="cover value"
      />
    </div>
    <aside
      class="otherBooks"
    >
      <h2>
        Other books by this author
      </h2>
      <div
        class="container"
      >
        <div
          class="container"
        >
          <div
            data-testid="book-list-items"
          >
            <div
              class="container"
            >
              <div
                class="details"
              >
                <h2
                  class="title"
                  data-testid="book-title"
                >
                  title value 2
                </h2>
                <div
                  class="author"
                >
                  by 
                  <span
                    data-testid="book-author"
                  >
                    author value
                  </span>
                </div>
                <div
                  class="link"
                  data-testid="book-link"
                >
                  <a
                    href="/%20value%202"
                  >
                    More Info
                  </a>
                </div>
              </div>
              <div
                class="cover"
              >
                <img
                  alt="title value 2 book cover"
                  class="coverImage"
                  data-testid="book-image"
                  src="cover value 2"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </aside>
  </div>
</DocumentFragment>
`;

exports[`<BookDetails /> renders as expected when loading 1`] = `
<DocumentFragment>
  <div
    class="container"
    data-testid="book-details"
  >
    <h1
      class="title"
      data-testid="title"
    >
      title value
    </h1>
    <div
      class="author"
    >
      by 
      <span
        data-testid="author"
      >
        author value
      </span>
    </div>
    <div
      class="isbn"
    >
      ISBN: 
      <span
        data-testid="isbn"
      >
        isbn value
      </span>
    </div>
    <div
      class="image"
    >
      <img
        alt="title value book cover"
        class="image"
        data-testid="image"
        src="cover value"
      />
    </div>
    <aside
      class="otherBooks"
    >
      <h2>
        Other books by this author
      </h2>
      <div
        class="container isLoading"
      >
        <div
          class="notification"
          data-testid="loader"
        >
          Loading...
        </div>
      </div>
    </aside>
  </div>
</DocumentFragment>
`;

exports[`<BookDetails /> renders content as expected 1`] = `
<DocumentFragment>
  <div
    class="container"
    data-testid="book-details"
  >
    <h1
      class="title"
      data-testid="title"
    >
      title value
    </h1>
    <div
      class="author"
    >
      by 
      <span
        data-testid="author"
      >
        author value
      </span>
    </div>
    <div
      class="isbn"
    >
      ISBN: 
      <span
        data-testid="isbn"
      >
        isbn value
      </span>
    </div>
    <div
      class="image"
    >
      <img
        alt="title value book cover"
        class="image"
        data-testid="image"
        src="cover value"
      />
    </div>
    <aside
      class="otherBooks"
    >
      <h2>
        Other books by this author
      </h2>
      <div
        class="container isLoading"
      >
        <div
          class="notification"
          data-testid="loader"
        >
          Loading...
        </div>
      </div>
    </aside>
  </div>
</DocumentFragment>
`;

exports[`<BookDetails /> renders content as expected when there is no ISBN 1`] = `
<DocumentFragment>
  <div
    class="container"
    data-testid="book-details"
  >
    <h1
      class="title"
      data-testid="title"
    >
      title value
    </h1>
    <div
      class="author"
    >
      by 
      <span
        data-testid="author"
      >
        author value
      </span>
    </div>
    <div
      class="isbn"
    >
      ISBN: 
      <span
        data-testid="isbn"
      >
        not available
      </span>
    </div>
    <div
      class="image"
    >
      <img
        alt="title value book cover"
        class="image"
        data-testid="image"
        src="cover value"
      />
    </div>
    <aside
      class="otherBooks"
    >
      <h2>
        Other books by this author
      </h2>
      <div
        class="container isLoading"
      >
        <div
          class="notification"
          data-testid="loader"
        >
          Loading...
        </div>
      </div>
    </aside>
  </div>
</DocumentFragment>
`;
